src/target/riscv/riscv.c

```diff
@@ -3625,14 +3628,40 @@ int riscv_get_register(struct target *target, riscv_reg_t *value,
 
 	int result = r->get_register(target, value, regid);
 
-	if (result == ERROR_OK)
+	if (result == ERROR_OK) {
+		/* Update the cache in case we're called from
+		 * riscv_save_register(). */
+		buf_set_u64(reg->value, 0, reg->size, *value);
 		reg->valid = gdb_regno_cacheable(regid, false);
+	}
 
 	LOG_DEBUG("[%s] %s: %" PRIx64, target_name(target),
 			gdb_regno_name(regid), *value);
 	return result;
 }
 
+int riscv_save_register(struct target *target, enum gdb_regno regid)
+{
+	riscv_reg_t value;
+	if (!target->reg_cache) {
+		assert(!target_was_examined(target));
+		return ERROR_OK;
+	}
+
+	struct reg *reg = &target->reg_cache->reg_list[regid];
+	LOG_DEBUG("[%s] save %s", target_name(target), reg->name);
+	if (riscv_get_register(target, &value, regid) != ERROR_OK)
+		return ERROR_FAIL;
+
```

Comment:
So, the variable named "value" never gets used after riscv_get_register call, because what we care about when calling riscv_get_register here is the side effect of priming the register value into the underlying cache.  Makes sense.  And I guess the compiler isn't flagging that as a warning, which is OK because "we know what we're doing" :)

Reply:
I need that on a t-shirt. :-)